### ISSUE TITLE
Status doesn't change when changing tmux windows

### ISSUE THREAD

**Issue by m1foley** [vimpostor/vim-tpipeline#2]
This is a really cool plugin! Please tell me if I'm missing something here.

# Steps to recreate

1. Open Vim. The tmux status line is set to the Vim statusline as expected.
2. Open a new tmux tab.

Expected: The second tmux window should have an empty status line.
Actual: The second tmux window still shows the Vim statusline from the first window.

A similar thing happens if I have Vim open in two windows: switching between them doesn't update the statusline. To update it I have to do something like quit & reopen Vim.

# Settings

My `.tmux.conf` is [here](https://github.com/m1foley/dotfiles/blob/main/_tmux.conf) but with the following changes:
```diff
+set -g focus-events on
-set -g status-interval 30
+set -g status-interval 5
-set -g status-right "#{battery_icon}#{battery_percentage} %F %H:%M"
+set -g status-right '#(cat #{socket_path}-\#{session_id}-vimbridge)'
+set -g status-right-length 120
+set -g status-interval 0
```

My `.vimrc` is [here](https://github.com/m1foley/dotfiles/blob/main/_vimrc) but with the following changes:
```diff
+Plug 'vimpostor/vim-tpipeline'
```

**Comment by vimpostor** [comment]
Both of those things are working for me.

I see you already have configured tmux correctly for focus events. What terminal emulator and what vim version are you using? Both of those need to support focus events as well.

To check if focus events are working, you can use `:autocmd FocusGained * echom 'Test'` and then unfocus and focus your vim to see if you see a test message.

**Comment by m1foley** [comment]
That `autocmd` is not producing any messages when I'm in terminal Vim! It does work when I use the gvim version of MacVim. A brief web search shows a project called [vim-tmux-focus-events](https://github.com/tmux-plugins/vim-tmux-focus-events) that says, "FocusGained and FocusLost autocommand events are not working in terminal vim." To get events working do I need to install a plugin like that?

 I'm using the Apple "Terminal" program (declared as `xterm-256color`).

```
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Dec 28 2020 12:55:00)
macOS version
Included patches: 1-2164
Compiled by Homebrew
Huge version with MacVim GUI.  Features included (+) or not (-):
+acl               +clientserver      +diff              +folding           +lambda            +mouseshape        +netbeans_intg     +python3           -sun_workshop      +timers            +wildignore
+arabic            +clipboard         +digraphs          -footer            +langmap           +mouse_dec         +num64             +quickfix          +syntax            +title             +wildmenu
+autocmd           +cmdline_compl     +dnd               +fork()            +libcall           -mouse_gpm         +odbeditor         +reltime           +tag_binary        +toolbar           +windows
+autochdir         +cmdline_hist      -ebcdic            +fullscreen        +linebreak         -mouse_jsbterm     +packages          +rightleft         -tag_old_static    +transparency      +writebackup
-autoservername    +cmdline_info      +emacs_tags        +gettext           +lispindent        +mouse_netterm     +path_extra        +ruby              -tag_any_white     +user_commands     -X11
+balloon_eval      +comments          +eval              -hangul_input      +listcmds          +mouse_sgr         +perl              +scrollbind        +tcl               +vartabs           -xfontset
+balloon_eval_term +conceal           +ex_extra          +iconv             +localmap          -mouse_sysmouse    +persistent_undo   +signs             +termguicolors     +vertsplit         +xim
+browse            +cryptv            +extra_search      +insert_expand     +lua               +mouse_urxvt       +popupwin          +smartindent       +terminal          +virtualedit       -xpm
++builtin_terms    +cscope            -farsi             +ipv6              +menu              +mouse_xterm       +postscript        -sound             +terminfo          +visual            -xsmp
+byte_offset       +cursorbind        +file_in_path      +job               +mksession         +multi_byte        +printer           +spell             +termresponse      +visualextra       -xterm_clipboard
+channel           +cursorshape       +find_in_path      +jumplist          +modify_fname      +multi_lang        +profile           +startuptime       +textobjects       +viminfo           -xterm_save
+cindent           +dialog_con_gui    +float             +keymap            +mouse             -mzscheme          -python            +statusline        +textprop          +vreplace
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "$VIM/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/Applications/MacVim.app/Contents/Resources/vim"
Compilation: clang -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_MACVIM -Wall -Wno-unknown-pragmas -pipe  -DMACOS_X -DMACOS_X_DARWIN  -fdeclspec -arch x86_64 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
Linking: clang   -L. -fstack-protector-strong -L/usr/local/lib -L/usr/local/opt/libyaml/lib -L/usr/local/opt/openssl@1.1/lib -L/usr/local/opt/readline/lib -L. -fstack-protector-strong -L/usr/local/lib -L/usr/l
ocal/opt/libyaml/lib -L/usr/local/opt/openssl@1.1/lib -L/usr/local/opt/readline/lib  -arch x86_64 -L/usr/local/lib -o Vim        -lm  -lncurses -liconv /usr/local/lib/libintl.a -framework AppKit  -L/usr/local/
opt/lua/lib -llua5.4 -fstack-protector  -L/System/Library/Perl/5.18/darwin-thread-multi-2level/CORE -lperl  -L/usr/local/opt/python@3.9/Frameworks/Python.framework/Versions/3.9/lib/python3.9/config-3.9-darwin
-lpython3.9 -framework CoreFoundation -F/System/Library/Frameworks -framework Tcl -framework CoreFoundation -lruby.3.0  -L/usr/local/Cellar/ruby/3.0.0/lib
```

```sh
$ tmux -V
tmux 3.1c
```

**Comment by vimpostor** [comment]
You do not need that other plugin. Vim has native support for focus events since version `8.2.2345`. It seems your vim version is slightly before that patch level, so if you are able to update your vim version, your problem will probably be fixed.
Alternatively you can use Neovim which supports focus events since a long time.

My best guess is that Apple "Terminal" probably has focus event support, but I am not 100% sure. Please do report if you are able to make it work with Apple "Terminal" if you are able to update your vim.

**Comment by m1foley** [comment]
Thanks. Looks like I didn't read the README close enough. The Homebrew version of MacVim needs to be updated. I'll be lazy and wait for that to happen, and will report back if that works for me in "Terminal".

**Comment by vimpostor** [comment]
Thanks, I should probably put the requirements more prominently in the README.

**Comment by vimpostor** [comment]
Yup I just had to investigate this myself, as I broke this in my own config by moving from `xterm-256color` to `tmux-256color`.

The tl;dr is that the world of terminal capabilities is an utter mess. So in order to tell applications that a terminal emulator supports focus events, it has the `XT` capability. Since you have focus events working outside of tmux, when you run `tput XT` you will get an `EXIT_SUCCESS` return code. Inside tmux when you run the same command, you get an error. The easiest way to fix this is to put `set -g default-terminal "xterm-256color"` in your tmux config, which will amongst other things set the XT capability. However, usage of `xterm-256color` is discouraged by tmux and they encourage you to use `tmux-256color`. But the problem is that `tmux-256color` does not export the XT capability. Usually one can override capabilites, for example to add the True color extension one can add `set -as terminal-overrides ",*:Tc"`, but this doesn't work with the XT capability (`set -as terminal-overrides ",*:Tc:XT"`) - probably because tmux only overrides the terminal on the outside.

I haven't found a way to let tmux set the XT capability inside tmux, I will have to investigate this some more. Until then you can temporarily use `xterm-256color` (I haven't found any problems with this). I will keep this issue open until I have found a proper way to add the XT capability inside tmux.

**Comment by m1foley** [comment]
Holy smokes. Thanks for shaving this yak!

**Comment by vimpostor** [comment]
So there are basically two options:
- Put `set -g default-terminal "xterm-256color"` in your tmux config like I said
- Put the following block in your vimrc
```vim
let &t_fe = "\<Esc>[?1004h"
let &t_fd = "\<Esc>[?1004l"
```

Also relevant: https://github.com/tmux/tmux/issues/2606
